### PR TITLE
createImageBitmap Abstraction

### PR DESCRIPTION
**Pull request by CedricGuillemet**
Followup for https://forum.babylonjs.com/t/babylon-native-on-ubuntu/15615/21
createImageBitmap is used for `CreateGroundFromHeightMap` . This PR is the JS part. Native part will be done once this is OK.

The goal here is also for me to learn how to properly wrap JS function as the same work will be needed for the Canvas/GUI.

### CODE REVIEW

**Review by deltakosh — APPROVED** [review]

File: src/Engines/engine.ts [hunk 1]
```diff
@@ -290,6 +290,17 @@ export class Engine extends ThinEngine {
...
 
+    /**
+     * Engine abstraction for createImageBitmap
+     */
+    public createImageBitmap(...args: any[]): Promise<ImageBitmap> //image: ImageBitmapSource, options?: ImageBitmapOptions
+    {
+        if (args.length <= 2) {
+            return createImageBitmap(args[0], args[1]);
+        }
+        return createImageBitmap(args[0], args[1], args[2], args[3], args[4], args[5]);
```
Comment (bghgary) [inline src/Engines/engine.ts:301]: I think you can probably use `createImageBitmap.apply` here?

File: src/Engines/nativeEngine.ts [hunk 2]
```diff
@@ -1771,6 +1772,20 @@ export class NativeEngine extends Engine {
         }
     }
 
+    /**
+     * Engine abstraction for createImageBitmap with Native implementation
+     */
+    public createImageBitmap(...args: any[]): Promise<ImageBitmap>
+    {
+        return new Promise((resolve, reject) => {
+            if (args.length <= 2) {
```
Comment (bghgary) [inline src/Engines/nativeEngine.ts:1781]: ditto on this

File: src/XR/features/WebXRImageTracking.ts [hunk 3]
```diff
@@ -190,7 +191,8 @@ export class WebXRImageTracking extends WebXRAbstractFeature {
                         img.src = image.src;
                         img.onload = () => {
                             img.decode().then(() => {
-                                createImageBitmap(img).then((imageBitmap) => {
+                                const engine = EngineStore.LastCreatedEngine;
```
Comment (bghgary) [inline src/XR/features/WebXRImageTracking.ts:194]: `this._xrSessionManager.scene.getEngine()`

File: src/Engines/webgpuEngine.ts [hunk 4]
```diff
@@ -2084,7 +2086,8 @@ export class WebGPUEngine extends Engine {
             gpuTextureWrapper = this._textureHelper.createGPUTextureForInternalTexture(texture);
         }
 
-        createImageBitmap(video).then((bitmap) => {
+        const engine = EngineStore.LastCreatedEngine;
+        engine && engine.createImageBitmap(video).then((bitmap) => {
```
Comment (bghgary) [inline src/Engines/webgpuEngine.ts:2090]: This is the engine class right?

```suggestion
        this.createImageBitmap(video).then((bitmap) => {
```

File: src/Engines/webgpuEngine.ts [hunk 5]
```diff
@@ -2030,7 +2031,8 @@ export class WebGPUEngine extends Engine {
             gpuTextureWrapper = this._textureHelper.createGPUTextureForInternalTexture(texture, width, height);
         }
 
-        createImageBitmap(canvas).then((bitmap) => {
+        const engine = EngineStore.LastCreatedEngine;
+        engine && engine.createImageBitmap(canvas).then((bitmap) => {
```
Comment (bghgary) [inline src/Engines/webgpuEngine.ts:2035]: ditto

Comment (Popov72) [reply src/Engines/webgpuEngine.ts:2035]: You should be able to simply do `this.createImageBitmap(...)`?

File: src/Engines/webgpuEngine.ts (hunk 4 above)
Comment (Popov72) [reply src/Engines/webgpuEngine.ts:2090]: ..ah, Gary already pointed to it!

File: src/Misc/fileTools.ts [hunk 6]
```diff
@@ -160,7 +160,7 @@ export class FileTools {
 
         if (typeof Image === "undefined" || (engine?._features.forceBitmapOverHTMLImageElement ?? false)) {
             FileTools.LoadFile(url, (data) => {
-                createImageBitmap(new Blob([data], { type: mimeType }), { premultiplyAlpha: "none" }).then((imgBmp) => {
+                engine?.createImageBitmap(new Blob([data], { type: mimeType }), { premultiplyAlpha: "none" }).then((imgBmp) => {
```
Comment (Popov72) [inline src/Misc/fileTools.ts:163]: You should probably do `engine!` instead, as else we will be stuck without any error message if engine is null/undefined (even if it's only theoritical).

File: src/Engines/engine.ts (hunk 1 above)
Comment (CedricGuillemet) [reply src/Engines/engine.ts:301]: Exactly what I was looking for!

**Comment by CedricGuillemet** [comment]
I had to add a new abstraction function `resizeImageBitmap` that resize/convert to RGBA32 and returns an uint8array. I'm not sure the function name is appropriate as it does more than just resizing. Let me know.

File: src/Engines/engine.ts [hunk 7]
```diff
@@ -290,6 +291,39 @@ export class Engine extends ThinEngine {
         return EngineStore.LastCreatedScene;
     }
 
+    /**
+     * Engine abstraction for createImageBitmap
+     * @param args parameters for createImageBitmap implementation
+     * @returns ImageBitmap
+     */
+    public createImageBitmap(...args: any[]): Promise<ImageBitmap>
```
Comment (ryantrem) [inline src/Engines/engine.ts:299]: Why does this have to be `any[]` for parameters? I think there is something fundamental about this API that I'm not understanding.

Comment (CedricGuillemet) [reply src/Engines/engine.ts:299]: createImageBitmap has 2 signatures : https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap
So, aby number of parameters are supported here.

Comment (bghgary) [reply src/Engines/engine.ts:299]: Hmm, having two signatures doesn't imply that you have to use any or `...`. You can just create overloads in TypeScript (as long as there are different number of arguments).

Comment (bghgary) [reply src/Engines/engine.ts:299]: From lib.dom.d.ts:

```typescript
declare function createImageBitmap(image: ImageBitmapSource, options?: ImageBitmapOptions): Promise<ImageBitmap>;
declare function createImageBitmap(image: ImageBitmapSource, sx: number, sy: number, sw: number, sh: number, options?: ImageBitmapOptions): Promise<ImageBitmap>;
```

File: src/Engines/nativeEngine.ts [hunk 8]
```diff
@@ -143,6 +143,8 @@ interface INativeEngine {
     setTextureAnisotropicLevel(texture: WebGLTexture, value: number): void;
     setTexture(uniform: WebGLUniformLocation, texture: Nullable<WebGLTexture>): void;
     deleteTexture(texture: Nullable<WebGLTexture>): void;
+    createImageBitmap(parameters: any): ImageBitmap;
```
Comment (bghgary) [inline src/Engines/nativeEngine.ts:146]: this doesn't need to be `any` anymore.

File: src/Engines/nativeEngine.ts [hunk 9]
```diff
@@ -1771,6 +1773,38 @@ export class NativeEngine extends Engine {
...
+     */
+    public createImageBitmap(...args: any[]): Promise<ImageBitmap>
+    {
+        return new Promise((resolve, reject) => {
+            if (args.length && Array.isArray(args[0]) && args[0].length) {
+                const image = this._native.createImageBitmap(args[0][0]);
+                if (image) {
+                    resolve(image);
+                } else if (reject) {
+                    reject();
```
Comment (bghgary) [inline src/Engines/nativeEngine.ts:1789]: should probably put a reason in the reject

File: src/Engines/nativeEngine.ts [hunk 10]
```diff
@@ -1771,6 +1773,38 @@ export class NativeEngine extends Engine {
...
+                }
+                return;
+            }
+            throw new Error(`Unsupported data for createImageBitmap.`);
```
Comment (bghgary) [inline src/Engines/nativeEngine.ts:1793]: this is a bit odd, although this will translate to a reject call... why not just call reject here?

**Comment by deltakosh** [comment]
What is the current status of this PR?

File: src/Engines/engine.ts (hunk 7 above)
Comment (CedricGuillemet) [reply src/Engines/engine.ts:299]: from what I understand, you can declare overloadings but you can't have different signature for the implementation. I case either check parameters types for choosing the correct overloading or use `any` with `apply`. Which one shall I do?

**Comment by CedricGuillemet** [comment]
@deltakosh still working on TS function signature.

File: src/Engines/engine.ts (hunk 7 above)
Comment (bghgary) [reply src/Engines/engine.ts:299]: https://stackoverflow.com/questions/12688275/is-there-a-way-to-do-method-overloading-in-typescript

Here is an example: https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD29oG8CwAoaXoAcCuARiAJbDQBCYATgBRgBc0AdngLYECmVANNAYy3ZcAlIwBu8YgBMA3Bmy5CJMpVoNoEAC5ViTAOajoE6XMzZ8RUhWp0BbTlWgAfDdt17eBAPx2hVQ8ZSKPLYAL4YoUA

Comment (CedricGuillemet) [reply src/Engines/engine.ts]: I was taking the problem from the wrong side. there is only 1 override used in whole engine. So I changed the signature to only support it.

**Review by bghgary — APPROVED** [review]

**Pull request merged**